blocks/modal/modal.js

```diff
@@ -0,0 +1,55 @@
+import { loadFragment } from '../fragment/fragment.js';
+import {
+  buildBlock, decorateBlock, decorateIcons, loadBlock, loadCSS,
+} from '../../scripts/aem.js';
+
+// This is not a traditional block, so there is no decorate function. Instead, links to
+// a */modals/* path  are automatically transformed into a modal. Other blocks can also use
+// the createModal() and openModal() functions.
+
+export async function createModal(contentNodes) {
+  await loadCSS(`${window.hlx.codeBasePath}/blocks/modal/modal.css`);
+  const dialog = document.createElement('dialog');
+  const dialogContent = document.createElement('div');
+  dialogContent.classList.add('modal-content');
+  dialogContent.append(...contentNodes);
+  dialog.append(dialogContent);
+
+  const closeButton = document.createElement('button');
+  closeButton.classList.add('close-button');
+  closeButton.setAttribute('aria-label', 'Close');
+  closeButton.type = 'button';
+  closeButton.innerHTML = '<span class="icon icon-close"></span>';
+  closeButton.addEventListener('click', () => dialog.close());
+  dialog.append(closeButton);
+
+  // close dialog on clicks outside the dialog. https://stackoverflow.com/a/70593278/79461
+  dialog.addEventListener('click', (event) => {
+    const dialogDimensions = dialog.getBoundingClientRect();
+    if (event.clientX < dialogDimensions.left || event.clientX > dialogDimensions.right
+      || event.clientY < dialogDimensions.top || event.clientY > dialogDimensions.bottom) {
+      dialog.close();
+    }
+  });
+
+  dialog.addEventListener('close', () => dialog.remove());
+
+  const block = buildBlock('modal', '');
+  document.querySelector('main').append(block);
+  decorateBlock(block);
+  await loadBlock(block);
+  decorateIcons(closeButton);
+
+  block.append(dialog);
+  return { block, showModal: () => dialog.showModal() };
```

Comment:
here's a weird one. when the modal has a scrollbar, should the content be scrolled to the top, mid, bottom? if a user scrolls in a modal, closes it, and reopens, should their scroll position be preserved or start over from the top?

Reply:
done, scrolling to the top now. `dialogContent.scrollTop = 0;`